logs_sdk/lib/opentelemetry/sdk/logs/logger_provider.rb

```diff
@@ -0,0 +1,127 @@
+      # {OpenTelemetry::Logs::LoggerProvider}.
+      class LoggerProvider < OpenTelemetry::Logs::LoggerProvider
+        attr_reader :resource
+
+        EMPTY_NAME_ERROR = 'LoggerProvider#logger called without '\
+            'providing a logger name.'
+
+        # Returns a new {LoggerProvider} instance.
+        #
+        # @param [optional Resource] resource The resource to associate with
+        #   new LogRecords created by Loggers created by this LoggerProvider.
+        # @param [optional Array] log_record_processors Log Record Processors to
+        #   associate with the LoggerProvider.
+        #
+        # @return [LoggerProvider]
+        def initialize(
+          resource: OpenTelemetry::SDK::Resources::Resource.create,
+          log_record_processors: []
+        )
+          @log_record_processors = log_record_processors
+          @mutex = Mutex.new
+          @resource = resource
+          @stopped = false
+        end
+
+        # Returns a {OpenTelemetry::SDK::Logs::Logger} instance.
+        #
+        # @param [optional String] name Instrumentation package name
+        # @param [optional String] version Instrumentation package version
+        #
+        # @return [OpenTelemetry::SDK::Logs::Logger]
+        def logger(name = nil, version = nil)
+          name ||= ''
+          version ||= ''
+
+          OpenTelemetry.logger.warn(EMPTY_NAME_ERROR) if name.empty?
+
+          @mutex.synchronize do
+            OpenTelemetry::SDK::Logs::Logger.new(name, version, self)
+          end
+        end
+
+        # Adds a new LogRecordProcessor to this {LoggerProvider}'s
+        # log_record_processors.
+        #
+        # @param [LogRecordProcessor] log_record_processor The new
+        #   LogRecordProcessor to add.
+        def add_log_record_processor(log_record_processor)
+          @mutex.synchronize do
+            @log_record_processors = @log_record_processors.dup.push(log_record_processor)
+          end
+        end
+
+        # Attempts to stop all the activity for this {LoggerProvider}. Calls
+        # LogRecordProcessor#shutdown for all registered LogRecordProcessors.
+        #
+        # This operation may block until all the Log Records are processed. Must
+        # be called before turning off the main application to ensure all data
+        # are processed and exported.
+        #
+        # After this is called all the newly created {LogRecord}s will be no-op.
+        #
+        # @param [optional Numeric] timeout An optional timeout in seconds.
+        # @return [Integer] Export::SUCCESS if no error occurred, Export::FAILURE if
+        #   a non-specific failure occurred, Export::TIMEOUT if a timeout occurred.
+        def shutdown(timeout: nil)
+          @mutex.synchronize do
+            if @stopped
+              OpenTelemetry.logger.warn('LoggerProvider#shutdown called multiple times.')
+              return OpenTelemetry::SDK::Logs::Export::FAILURE
+            end
```

Comment:
Why does `#shutdown` return a failure status code if called when `@stopped` is true, but `#force_flush` returns a success status code?

Reply:
Also, the spec states: 
>After the call to Shutdown, subsequent attempts to get a Logger are not allowed. SDKs SHOULD return a valid no-op Logger for these calls, if possible.
[Source](https://opentelemetry.io/docs/specs/otel/logs/sdk/#shutdown)

This is also true for `TracerProvider#shutdown` [(source)](https://opentelemetry.io/docs/specs/otel/trace/sdk/#shutdown). Should I update this code in the Logs SDK to return `OpenTelemetry::Logs::LoggerProvider::NOOP_LOGGER`? If so, should action be taken on the Trace SDK?